source/extensions/quic_listeners/quiche/quic_filter_manager_connection_impl.h

```diff
@@ -0,0 +1,129 @@
+#pragma once
+
+#include "envoy/event/dispatcher.h"
+#include "envoy/network/connection.h"
+
+#include "common/common/empty_string.h"
+#include "common/common/logger.h"
+#include "common/network/filter_manager_impl.h"
+#include "common/stream_info/stream_info_impl.h"
+
+#include "extensions/quic_listeners/quiche/envoy_quic_connection.h"
+
+namespace Envoy {
+namespace Quic {
+
+// Act as a Network::Connection to HCM and a FilterManager to FilterFactoryCb.
+class QuicFilterManagerConnectionImpl : public Network::FilterManagerConnection,
+                                        protected Logger::Loggable<Logger::Id::connection> {
+public:
+  QuicFilterManagerConnectionImpl(std::unique_ptr<EnvoyQuicConnection> connection,
+                                  Event::Dispatcher& dispatcher);
+
+  // Network::FilterManager
+  // Overridden to delegate calls to filter_manager_.
+  void addWriteFilter(Network::WriteFilterSharedPtr filter) override;
+  void addFilter(Network::FilterSharedPtr filter) override;
+  void addReadFilter(Network::ReadFilterSharedPtr filter) override;
+  bool initializeReadFilters() override;
+
+  // Network::Connection
+  void addConnectionCallbacks(Network::ConnectionCallbacks& cb) override;
+  void addBytesSentCallback(Network::Connection::BytesSentCb /*cb*/) override {
+    // TODO(danzh): implement to support proxy. This interface is only called from
+    // TCP proxy code.
+    NOT_REACHED_GCOVR_EXCL_LINE;
+  }
+  void enableHalfClose(bool enabled) override;
+  void close(Network::ConnectionCloseType type) override;
+  Event::Dispatcher& dispatcher() override { return dispatcher_; }
+  uint64_t id() const override { return id_; }
+  std::string nextProtocol() const override { return EMPTY_STRING; }
+  void noDelay(bool /*enable*/) override {
+    // No-op. TCP_NODELAY doesn't apply to UDP.
+  }
+  void setDelayedCloseTimeout(std::chrono::milliseconds timeout) override;
+  std::chrono::milliseconds delayedCloseTimeout() const override;
+  void readDisable(bool disable) override {
+    ASSERT(!disable, "Quic connection should be able to read through out its life time.");
```

Comment:
For these normal asserts, do these functions get called? Or should they not get called at all? Same elsewhere.

Reply:
It shouldn't be called for multiplexing case according to my reading of the HCM implementation. I might be wrong, but NOT_REACHED for now.